Add Buildah as an alternative container image build tool

azatsarynnyy opened the pull request:
In this PR I propose to add [Buildah](https://github.com/containers/buildah) support as an alternative container image build tool that is used for basic image operations by `Kubernetes: Run` and `Kubernetes: Debug` commands.

By default, Docker is used. And if the user prefers using Buildah it can be enabled with `preferBuildah` configuration.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/vscode-kubernetes-tools?pullRequest=559) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/vscode-kubernetes-tools?pullRequest=559) before we can accept your contribution.<br/><br/>:x: azatsarynnyy [sign now](https://cla.opensource.microsoft.com/Azure/vscode-kubernetes-tools?pullRequest=559)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/vscode-kubernetes-tools?pullRequest=559) it.</sub>

azatsarynnyy commented:
cc @gorkem

itowlson commented on src/imageBuild/buildToolsRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+export let currentBuildTool: string;
+
+/** Starts listening changing the preferred build tool. */
+export function start(): vscode.Disposable {
```
What would be the cost of just looking up the config entry each time, instead of having a mutable variable and an event listener?  We do this throughout the extension and it feels simpler and doesn't seem to cause performance problems (he said cautiously, acutely aware that we _do_ have performance problems).

itowlson replied:
Oh, it's because it's not just looking in config, it's also checking if `buildah` is present.  And we definitely _don't_ want to do that every time!

itowlson replied:
For most tools we defer detection until the user tries to use the tool.  I think we check for `kubectl` on startup but that is just to warn - we don't rely on it for a setting.  We also have existing mechanisms for checking if a binary is present (see e.g. `binutil.ts`) which might be faster than calling the executable.

itowlson commented on src/imageBuild/buildToolsRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+
+/** Starts listening changing the preferred build tool. */
+export function start(): vscode.Disposable {
+    detectCurrentBuildTool();
```
This makes me anxious... in practice it should never be an issue, but in theory `detectCurrentBuildTool` could take a long time to run on extension startup, and so `currentBuildTool` could be accessed before it had been set.

itowlson commented on package.json:
```diff
@@ -231,6 +231,10 @@
                             "type": "boolean",
                             "description": "If true, show Kubernetes resource commands on file context menu for all YAML files"
+                        },
+                        "preferBuildah": {
```
I am not sure this is the right design here - what if someone comes along with a PR to support another build tool, say `img`?  How would we manage the interaction between `preferBuildah` and a hypothetical `preferImg`?

Suggest instead a `preferredImageBuildTool` with an enum that restricts it to be `Docker` or `Buildah` (cf. the `kubectlVersioning` config setting).

itowlson commented on src/imageBuild/buildToolsRegistry.ts:
```diff
@@ -0,0 +1,94 @@
+import { getPreferBuildah } from '../components/config/config';
+import { shell } from '../shell';
+
+interface ContainerImageBuildTool {
```
I like the way you've done this - should make it easy if people want to add other build tools in future.

azatsarynnyy commented:
Hello @itowlson, thank you for your review!
I've addressed all your notes and updated the PR:
- reworked the config design to the list of the possible tools;
- replaced a tool detection on the extension start with the existing mechanisms provided by `binutil.ts`.

itowlson commented:
Could you check the status of your CLA and sign if you haven't already done so please?  Unfortunately we can't take this PR without a CLA.

https://github.com/Azure/vscode-kubernetes-tools/pull/559#issuecomment-499184283

itowlson commented:
Thank you @azatsarynnyy for the CLA sign-off - we have lift-off!
